From 95f111400889d62b22ef764b171ef7dcce90fe13 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Mon, 8 Aug 2016 15:52:48 +0300
Subject: [PATCH 0680/2241] clk: apn806: Add eMMC clock to system controller
 driver

- Add fixed clock of 400MHz to system controller driver
  This clock is used as SD/eMMC clock source

Change-Id: Ifde0926493e0416d326fa4ad712169f8d0b28516
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31908
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 arch/arm64/boot/dts/marvell/armada-ap806.dtsi | 2 +-
 drivers/clk/mvebu/ap806-system-controller.c   | 8 +++++++-
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/boot/dts/marvell/armada-ap806.dtsi b/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
index 5991197..bae6d6e 100644
--- a/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
@@ -293,7 +293,7 @@
 				compatible = "marvell,ap806-system-controller", "syscon";
 				#clock-cells = <1>;
 				clock-output-names = "ap-cpu-cluster-0", "ap-cpu-cluster-1",
-						     "ap-fixed", "ap-mss";
+						     "ap-fixed", "ap-mss", "ap-emmc";
 				reg = <0x6f4000 0x1000>;
 			};
 
diff --git a/drivers/clk/mvebu/ap806-system-controller.c b/drivers/clk/mvebu/ap806-system-controller.c
index 7d9e892..7a2efde 100644
--- a/drivers/clk/mvebu/ap806-system-controller.c
+++ b/drivers/clk/mvebu/ap806-system-controller.c
@@ -24,7 +24,7 @@
 #define AP806_SAR_REG			0x400
 #define AP806_SAR_CLKFREQ_MODE_MASK	0x1f
 
-#define AP806_CLK_NUM			4
+#define AP806_CLK_NUM			5
 
 static struct clk *ap806_clks[AP806_CLK_NUM];
 
@@ -111,6 +111,12 @@ static void __init ap806_syscon_clk_init(struct device_node *np)
 	ap806_clks[3] = clk_register_fixed_factor(NULL, name, fixedclk_name,
 						  0, 1, 6);
 
+	/* eMMC Clock is fixed clock divided by 3 */
+	of_property_read_string_index(np, "clock-output-names",
+				      4, &name);
+	ap806_clks[4] = clk_register_fixed_factor(NULL, name, fixedclk_name,
+						  0, 1, 3);
+
 	of_clk_add_provider(np, of_clk_src_onecell_get, &ap806_clk_data);
 }
 
-- 
2.7.4

